components/brave_wallet/browser/zcash/rust/lib.rs

```diff
@@ -0,0 +1,146 @@
+    keys::FullViewingKey as OrchardFVK,
+    zip32::Error as Zip32Error,
+    zip32::ExtendedSpendingKey};
+
+use brave_wallet::{
+  impl_error
+};
+
+// The rest of the wallet code should be updated to use this version of unwrap
+// and then this code can be removed
+#[macro_export]
+macro_rules! impl_result {
+    ($t:ident, $r:ident, $f:ident) => {
+        impl $r {
+            fn error_message(self: &$r) -> String {
+                match &self.0 {
+                    Err(e) => e.to_string(),
+                    Ok(_) => "".to_string(),
+                }
+            }
+
+            fn is_ok(self: &$r) -> bool {
+                match &self.0 {
+                    Err(_) => false,
+                    Ok(_) => true,
+                }
+            }
+
+            // Unfortunately cxx doesn't support passing $r by value here so
+            // we have to clone the inner value instead of passing ownership
+            // This is not really a big deal because eventually we want to
+            // replace this with mojo which would serialize this anyway
+            fn unwrap(self: &$r) -> Box<$t> {
+                Box::new(self.0.as_ref().expect(
+                    "Unhandled error before unwrap call").clone())
+            }
+        }
+
+        impl From<Result<$f, Error>> for $r {
+            fn from(result: Result<$f, Error>) -> Self {
+                match result {
+                    Ok(v) => Self(Ok($t(v))),
+                    Err(e) => Self(Err(e)),
+                }
+            }
+        }
+    };
+}
+
+#[allow(unsafe_op_in_unsafe_fn)]
+#[cxx::bridge(namespace = brave_wallet::orchard)]
+mod ffi {
+    extern "Rust" {
+        type OrchardExtendedSpendingKey;
+
+        type OrchardExtendedSpendingKeyResult;
+
+        fn generate_orchard_extended_spending_key_from_seed(
+            bytes: &[u8]
+        ) -> Box<OrchardExtendedSpendingKeyResult>;
+
+        fn is_ok(self: &OrchardExtendedSpendingKeyResult) -> bool;
+        fn error_message(self: &OrchardExtendedSpendingKeyResult) -> String;
+        fn unwrap(self: &OrchardExtendedSpendingKeyResult) -> Box<OrchardExtendedSpendingKey>;
+
+        fn derive(
+            self: &OrchardExtendedSpendingKey,
+            index: u32
+        ) -> Box<OrchardExtendedSpendingKeyResult>;
+        fn external_address(
+            self: &OrchardExtendedSpendingKey,
+            diversifier_index: u32
+        ) -> [u8; 43];  // Array size should match kOrchardRawBytesSize
+        fn internal_address(
+            self: &OrchardExtendedSpendingKey,
+            diversifier_index: u32
+        ) -> [u8; 43];  // Array size should match kOrchardRawBytesSize
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Zip32(Zip32Error)
+}
+
+impl_error!(Zip32Error, Zip32);
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self {
+            Error::Zip32(e) => write!(f, "Error: {}", e.to_string()),
```

Comment:
nit: orchard implements the Display trait for `Zip32Error`, so this can just be
```rust
write!(f, "Zip32 Error: {e}"),
```
No need to call `to_string()`.